crates/binjs_io/src/entropy/read/prelude_decoders.rs

```diff
@@ -0,0 +1,299 @@
+/// it consumes the bytes `b"[[section2]]"`).
+pub struct SectionDecoder<T>
+where
+    T: Read,
+{
+    /// The latest name read.
+    name: Name<NameData>,
+
+    /// The input from which data will be read.
+    input: T,
+
+    /// The next pending char.
+    next_char: [u8; 1],
+}
+impl<T> SectionDecoder<T>
+where
+    T: Read,
+{
+    /// Create a new SectionDecoder to parse a given input.
+    pub fn new(input: T) -> Self {
+        Self {
+            name: Name::Nothing,
+            input,
+            next_char: [b'?'], // Arbitrary char.
+        }
+    }
+
+    /// The latest name read. This is either the name of a stream
+    /// or the name of the next section.
+    ///
+    /// This method returns
+    ///
+    /// - `Name::Nothing` if we have not encountered a name yet;
+    /// - `Name::Stream(&'foobar')` if the latest name we have encountered is `[foobar]`
+    ///    (single-brackets indicate a stream name)
+    /// - `Name::Section(&'snafu')` if the latest name we have encountered is `[[snafu]]`
+    ///    (double-brackets indicate a section name).
+    pub fn name(&self) -> Name<&[u8]> {
+        match self.name {
+            Name::Nothing => Name::Nothing,
+            Name::Stream(ref s) => Name::Stream(s.as_slice()),
+            Name::Section(ref s) => Name::Section(s.as_slice()),
+        }
+    }
+
+    /// Finalize this SectionReader and recover the input.
+    pub fn into_input(self) -> T {
+        self.input
+    }
+
+    /// Continue reading a name.
+    ///
+    /// This method is designed to be called from `read_stream`, after `read_stream`
+    /// has read the `[` that precedes the name.
+    fn read_name(&mut self) -> Result<NameData, TokenReaderError> {
+        if self.next_char[0] == b'[' {
+            self.input
+                .read_exact(&mut self.next_char)
+                .map_err(TokenReaderError::ReadError)?;
+        }
+        let mut name = NameData::new();
+        while self.next_char[0] != b']' && name.len() < NAME_MAX_LEN {
+            name.push(self.next_char[0]);
+            self.input
+                .read_exact(&mut self.next_char)
+                .map_err(TokenReaderError::ReadError)?;
+        }
+        Ok(name)
+    }
+
+    /// Read the next stream.
+    ///
+    /// A stream starts with a name `[foo_bar]`, followed by a number of compressed
+    /// bytes, followed by the actual bytes. In case of success, this returns the
+    /// buffer containing the decompressed bytes. To access the name of the stream,
+    /// use `self.name()`.
+    fn read_stream(&mut self) -> Result<Option<Vec<u8>>, TokenReaderError> {
+        debug!(target: "read", "SectionDecoder::read_stream");
+        self.input
+            .expect(b"[")
+            .map_err(|e| TokenReaderError::ReadError(e))?;
+
+        self.input
+            .read_exact(&mut self.next_char)
+            .map_err(TokenReaderError::ReadError)?;
+
+        if self.next_char[0] == b'[' {
```

Comment:
> I feel variable-length marker isn't good for performance, and also complicates the implementation unnecessarily.
>
> can it be something like PNG's chunk name, that is 4-bytes fixed?
> http://www.libpng.org/pub/png/spec/1.2/PNG-Structure.html

If you're talking about the number-of-bytes marker at the start of each stream, there are only ~12 instances of this marker during a complete decompression, so I'm not too worried about the performance cost.

I also don't really feel that it complicates the implementation – if we move to a 4-bytes encoding, we also need to specify/implement its endianness, which doesn't strike me as particularly simpler.

Fwiw, wasm uses a varuint, too.

> I guess we can use the similar rule for uppercase/lowercase, for denoting it's section or stream.
> or just use extra 4 bytes marker for "end of section".
>
> (the latter might be simpler, given the name of the next section isn't consumed by the previous section reader)
> 
> that way you don't have to check the existence of extra '[', or read bytes one-by-one looking for the next ']',
> but just reading 4-bytes finishes everything.

So, you suggest I add 4 bytes for "end of section", and get rid of section names entirely? A bit harder to debug, but a bit simpler to implement, so it makes sense.

On the other hand, we already know that we'll want new section types, for lazy functions/streaming, for licenses, possibly something for sourcemaps, etc. so I'd like to keep the code extensible.

For what it's worth, wasm also:

- 1 byte that may either be 0 (for custom sections) or non-0 (for standard sections);
- if the byte is 0, it's followed by a varuint string length, then a string payload.

We could uniformize with wasm.

> if the number of section/stream variant is smaller or even fixed, we won't need 4-bytes, but 1 byte might work.
> 
> also, wasm format will also have some hint
> https://github.com/WebAssembly/design/blob/master/BinaryEncoding.md
> 
> (I uses PNG format as example just because I'm more familiar with it than wasm)